File: frontend_flutter/lib/models/performance_metrics.dart
class PerformanceMetrics {
  final double fpsActual;
  final int framesProcessed;
  final int framesDropped;
  final double queuePressure;
  final bool adaptiveDegraded;
  final Map<String, dynamic> stageTiming;

  const PerformanceMetrics({
    required this.fpsActual,
    required this.framesProcessed,
    required this.framesDropped,
    required this.queuePressure,
    required this.adaptiveDegraded,
    required this.stageTiming,
  });

  factory PerformanceMetrics.fromJson(Map<String, dynamic> json) {
    return PerformanceMetrics(
      fpsActual: (json['fps_actual'] as num?)?.toDouble() ?? 0,
      framesProcessed: json['frames_processed'] as int? ?? 0,
      framesDropped: json['frames_dropped'] as int? ?? 0,
      queuePressure: (json['queue_pressure'] as num?)?.toDouble() ?? 0,
      adaptiveDegraded: json['adaptive_degraded'] as bool? ?? false,
      stageTiming: (json['stage_timing'] as Map?)?.cast<String, dynamic>() ?? {},
    );
  }
}

File: backend_python/app/models/performance_metrics.py
from __future__ import annotations

from pydantic import BaseModel


class StageTiming(BaseModel):
    capture_ms: float = 0.0
    stt_ms: float = 0.0
    ollama_ms: float = 0.0
    tts_ms: float = 0.0
    lipsync_ms: float = 0.0
    encode_ms: float = 0.0
    obs_ms: float = 0.0


class PerformanceMetrics(BaseModel):
    fps_actual: float = 0.0
    frames_processed: int = 0
    frames_dropped: int = 0
    queue_pressure: float = 0.0
    adaptive_degraded: bool = False
    stage_timing: StageTiming = StageTiming()
